### PR TITLE
fix(forms): collapse codeBasePath/customFunctionsPath overlap so custom functions resolve (FORMS-26373)

### PR DESCRIPTION
**Pull request by arunattri26**
## Problem
[FORMS-26373](https://jira.corp.adobe.com/browse/FORMS-26373) — custom functions fail to resolve when an EDS project is mounted **under a directory**.

In that setup `window.hlx.codeBasePath` ends with the mount folder (e.g. `…/eds-unified-cc`) and authors write `customFunctionsPath` from **repo root** (e.g. `/eds-unified-cc/blocks/form/functions.js`). The old runtime join was a raw concat:

```js
import(`${codeBasePath}${customFunctionsPath}`)
// /content/…/eds-unified-cc  +  /eds-unified-cc/blocks/form/functions.js
// → /content/…/eds-unified-cc/eds-unified-cc/blocks/form/functions.js   ← duplicated mount folder → 404
```

Meanwhile the AEM Rule Editor builds `base + customFunctionsPath` (repo root) and resolves it — producing the reported mutually-exclusive symptom: the functions load in the **rule editor** XOR at **runtime**, depending on whether the mount folder is included in the path.

## Fix
Add a single shared `resolveFunctionUrl(codeBasePath, path)` and use it for **both** the `modulepreload` hint (`preloadFunctionScripts`) and the actual dynamic `import()` (`registerCustomFunctions`), so the preloaded and imported URLs can never diverge.

`resolveFunctionUrl` joins the two fragments and **collapses the longest overlap where the tail of `codeBasePath` repeats the head of `path`**, matched at **segment boundaries** (so `/forms` ≠ `/forms-utils`) and spanning **one or more** segments. An absolute origin (`scheme://host`) is preserved.

```
resolveFunctionUrl('/content/…/eds-unified-cc', '/eds-unified-cc/blocks/form/functions.js')
  → /content/…/eds-unified-cc/blocks/form/functions.js   ✅
```

Authors write `customFunctionsPath` from repo root (what the rule editor already needs); the runtime de-dups the overlap. No backend / rule-editor change required.

## Compatibility
- **Root sites** (`codeBasePath === ''`) → path used as-is. Unchanged.
- **Legacy** `codeBasePath`-relative paths (no overlap) → simple join. Unchanged.
- **Mounted-under-a-directory** repo-root paths → overlap collapsed. Fixed.

## Tests
Adds `test/unit/functionRegistration.test.js` for `resolveFunctionUrl`: slash normalization, empty/missing base, no-overlap (legacy), single-segment overlap (exact HDFC values), multi-segment overlap, segment-boundary matching, and absolute-origin preservation. Full `mocha` suite green (256 passing); ESLint clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: blocks/form/rules/functionRegistration.js [hunk 1]
```diff
@@ -21,6 +21,47 @@ import { registerFunctions } from './model/afb-runtime.min.js';
...
+ *   resolveFunctionUrl('/a/b/eds-cc', '/eds-cc/blocks/form/functions.js')
+ *     -> '/a/b/eds-cc/blocks/form/functions.js'
+ * An absolute origin in codeBasePath (e.g. https://host) is preserved as-is. Used for
+ * both the modulepreload hint and the actual dynamic import so the two can never
+ * resolve to different URLs.
+ * @param {string} [codeBasePath] - e.g. window.hlx?.codeBasePath
+ * @param {string} path - path to join, with or without a leading slash
+ * @returns {string} the joined URL
+ */
+export function resolveFunctionUrl(codeBasePath, path) {
```
Comment (vdua) [inline blocks/form/rules/functionRegistration.js:38]:
```
 resolveFunctionUrl('../../', 'blocks/form/rules/functions.js')
  // actual:   /../../blocks/form/rules/functions.js
  // expected: ../../blocks/form/rules/functions.js
```

```
resolveFunctionUrl(
    '/content/site/blocks',
    'blocks/form/rules/functions.js',
  )
  // actual:   /content/site/blocks/form/rules/functions.js
  // expected: /content/site/blocks/blocks/form/rules/functions.js
```

Some failures. Check if this values of codebasePath are possible 

also add the tests for the actual function that it impacts as well (if none exists)
`preloadFunctionScripts`

Comment (arunattri26) [reply blocks/form/rules/functionRegistration.js:38]: Both repro'd.

1. codeBasePath ideally won't be relative since it is the path from where eds code is being loaded, default is repo root but if we have different code for different forms organized in folders then codeBasePath will ve upto that folder. But I have added handling for relative path anyways. Now only defaults to root-relative when `codeBasePath` is empty or already starts with `/`.

2. Not a bug — `'blocks/form/rules/functions.js'` is the hardcoded OOTB path (line 74), not an arbitrary string. Collapsing is intentionally uniform across both call sites rather than special-cased for OOTB; pinned as a test. Exposure is narrow — it'd require an EDS repo literally named `blocks`.

Also added tests for `preloadFunctionScripts` (href building, dedup) and `registerCustomFunctions` (previously untested), exercising the real dynamic import.

**Review by vdua — APPROVED** [review]
